ts/query.go

```diff
@@ -0,0 +1,431 @@
+// first offset for which *any* iterator in the set has real data.
+func (is iteratorSet) init() {
+	heap.Init(&is)
+	if !is.isValid() {
+		return
+	}
+	if is[0].nextReal.offset > 0 {
+		is.advance()
+	}
+}
+
+// advance advances each iterator in the set to the next value for which *any*
+// interpolatingIterator has a real value.
+func (is iteratorSet) advance() {
+	if !is.isValid() {
+		return
+	}
+
+	// All iterators in the set currently point to the same offset. Advancement
+	// begins by pre-advancing any iterators that have a real value for the
+	// current offset.
+	current := is[0].offset
+	for is[0].offset == current {
+		is[0].advanceTo(current + 1)
+		heap.Fix(&is, 0)
+	}
+
+	// It is possible that all iterators are now invalid.
+	if !is.isValid() {
+		return
+	}
+
+	// The iterator in position zero now has the lowest value for
+	// nextReal.offset - advance all iterators to that offset.
+	min := is[0].nextReal.offset
+	for i := range is {
+		is[i].advanceTo(min)
+	}
+	heap.Init(&is)
+}
+
+// timestamp returns a timestamp for the current offset of the iterators in this
+// set.
+func (is iteratorSet) timestamp() int64 {
+	if !is.isValid() {
+		return 0
+	}
+	return is[0].nextReal.timestampForOffset(is[0].offset)
+}
+
+// sumAvg returns the sum of the averages of all iterators in the set.
+func (is iteratorSet) sumAvg() float64 {
+	var sum float64
+	for i := range is {
+		sum += is[i].avg()
+	}
+	return sum
+}
+
+// Query returns data for the named time series during the supplied time span.
+// Data is returned as a series of consecutive data points.
+//
+// Data is queried only at the Resolution supplied: if data for the named time
+// series is not stored at the given resolution, an empty result will be
+// returned.
+//
+// All data stored on the server is downsampled to some degree; the data points
+// returned represent the average value within a sample period. Each datapoint's
+// timestamp falls in the middle of the sample period it represents.
+//
+// If data for a time series was collected from multiple sources, datapoints
+// will represent the sum across all sources at the same sample period. The
+// response will contain a list of all sources which were summed to produce the
+// result.
+func (db *DB) Query(name string, r Resolution, startNanos, endNanos int64) (
+	*proto.TimeSeriesQueryResult, error) {
+	// Normalize startNanos to the nearest SampleDuration boundary.
+	startNanos -= startNanos % r.SampleDuration()
+
+	// Based on the supplied timestamps and resolution, construct start and end
+	// keys for a scan that will return every key with data relevant to the
+	// query.
+	startKey := MakeDataKey(name, "", r, startNanos)
+	endKey := MakeDataKey(name, "", r, endNanos).Next()
+	scan := client.Scan(startKey, endKey, 0)
+	if err := db.kv.Run(scan); err != nil {
+		return nil, err
+	}
+	scanResponse := scan.Reply.(*proto.ScanResponse)
+
+	// Construct a new dataSpan for each distinct source encountered in the
```

Comment:
I'm actually just gonna read your `ts/doc.go`, if I feel something's missing after that I'll holler again.